### PR TITLE
Plot difference checkbox added to muon guis

### LINKED ISSUE #29070: Muon plot diff
We want to add a plot diff option to the plotting window in muon analysis. It should only be visible when in fitting mode.

### PR DESCRIPTION
**Pull request by tolu28-coder**
**Description of work.**
Added a plot difference checkbox to muon guis that displays or removes plot difference curve when clicked

**To test:**
1. Open muon analysis gui and load appropriate data.
2. Go to fitting tab add a fitting function and click on fit button.
3. Click on "Plot Difference" checkbox on top right of plotting window then click on fit. Plot difference curve should now appear.
4. Open frequency domain analysis gui and load appropriate data.
5. go to transform tab and click on calculate FTT.
6. Repeat steps 2 and 3.

Fixes #29070 . 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: scripts/Muon/GUI/Common/plot_widget/plot_widget_presenter.py [hunk 1]
```diff
@@ -116,11 +116,13 @@ def handle_plot_mode_changed_for_muon_analysis(self, plot_mode : PlotMode):
         self._view.set_plot_mode(str(plot_mode))
         if plot_mode == PlotMode.Data:
             self._view.enable_plot_type_combo()
+            self._view.plot_diff_checkbox.setVisible(False)
             self.update_plot()
             self.fitting_plot_range = self._figure_presenter.get_plot_x_range()
             self._figure_presenter.set_plot_range(self.data_plot_range)
         elif plot_mode == PlotMode.Fitting:
             self._view.disable_plot_type_combo()
+            self._view.plot_diff_checkbox.setVisible(True)
```
Comment (StephenSmith25) [inline scripts/Muon/GUI/Common/plot_widget/plot_widget_presenter.py:125]: Sorry I should have raised this before, but its only a minor issue. But can we implement a method similar to
disable_plot_type_combo and enable_plot_type_combo, within the [view ](https://github.com/mantidproject/mantid/blob/69a834894f26b22df9d14d5ef77024e934dc31c6/scripts/Muon/GUI/Common/plot_widget/plot_widget_view.py#L25) and the [view interface.](https://github.com/mantidproject/mantid/blob/69a834894f26b22df9d14d5ef77024e934dc31c6/scripts/Muon/GUI/Common/plot_widget/plot_widget_view_interface.py) The presenter here doesn't need to know that its dealing with a qt checkbox. Then we can replace the calls here with the new method.

File: scripts/test/Muon/plot_widget/plot_widget_presenter_test.py [hunk 2]
```diff
@@ -250,6 +250,8 @@ def test_handle_plot_selected_fits_correctly_calls_plot(self):
         fit = FitInformation(mock.MagicMock(), 'GaussOsc',
                              ['MUSR62260; Group; bottom; Asymmetry; MA'],
                              ['MUSR62260; Group; bottom; Asymmetry; MA; Fitted'])
+        self.view.is_plot_diff = mock.Mock
```
Comment (StephenSmith25) [inline scripts/test/Muon/plot_widget/plot_widget_presenter_test.py:253]: If we add a `is_plot_diff` method to the [view interface ](https://github.com/mantidproject/mantid/blob/69a834894f26b22df9d14d5ef77024e934dc31c6/scripts/Muon/GUI/Common/plot_widget/plot_widget_view_interface.py) we won't need to create the mock here.

Additionally, we should add a unit test here to check that the correct value is apssed to the model. We can also add a test to the model (plot_widget_model_test.py), to test the `get_fit_workspaces_and_indices` method.

File: docs/source/release/v5.1.0/muon.rst [hunk 3]
```diff
@@ -32,6 +32,7 @@ New Features
 - Addition of a LoadMuonNexusV2 algorithm to load the new Muon V2 files, see :ref:`LoadMuonNexusV2 <algm-LoadMuonNexusV2>`.
 - Updated rounding for time zero and first good data to be 3 decimal places.
 - Added double pulse analysis, see :ref:`Muon home tab <muon_home_tab-ref>`.
+- Added a Plot difference checkbox which users can use to pick if plot difference curve should be visible.
```
Comment (StephenSmith25) [inline docs/source/release/v5.1.0/muon.rst:35]:
```suggestion
- Added a Plot difference checkbox to the Muon Analysis GUI, which allows user to choose whether the fit difference curve is shown.
```

File: scripts/Muon/GUI/Common/plot_widget/plot_widget_view.py [hunk 4]
```diff
@@ -22,6 +22,7 @@ def warning_popup(message):
     def __init__(self, parent=None):
         super().__init__(parent=parent)
         self.setupUi(self)
+        self.plot_diff_checkbox.setVisible(False)
```
Comment (StephenSmith25) [inline scripts/Muon/GUI/Common/plot_widget/plot_widget_view.py:25]: We should set this somewhere in the `__init__` of the presenter using an `enable_plot_diff method` as mentioned above.

**Review by StephenSmith25 — COMMENTED** [review]
This works nicely for both the fitting and sequential fitting tabs, good work. Just some minor comments.

**Comment by AnthonyLim23** [comment]
@tolu28-coder you have conflicts

File: scripts/Muon/GUI/Common/plot_widget/plot_widget_view_interface.py [hunk 5]
```diff
@@ -17,6 +17,16 @@ class PlottingWidgetViewMeta(type(QtWidgets.QWidget), ABCMeta):
 
 class PlotWidgetViewInterface(metaclass=PlottingWidgetViewMeta):
 
+    @abstractmethod
+    def show_plot_diff(self):
+        #Shows Plot Difference Checkbox
```
Comment (StephenSmith25) [inline scripts/Muon/GUI/Common/plot_widget/plot_widget_view_interface.py:22]: These should be in the doc string format using Triple quotes.

See the other functions in this file

File: scripts/Muon/GUI/Common/plot_widget/plot_widget_model.py [hunk 6]
```diff
@@ -146,7 +146,7 @@ def _generate_run_indices(workspace_list):
         return indices
 
     @staticmethod
-    def get_fit_workspace_and_indices(fit):
+    def get_fit_workspace_and_indices(fit,plot_diff = True):
```
Comment (StephenSmith25) [inline scripts/Muon/GUI/Common/plot_widget/plot_widget_model.py:149]: Maybe change the argument name to `with_diff` or something along those lines - as `plot_diff` doesn't have much context here.

File: scripts/Muon/GUI/Common/plot_widget/plot_widget_model.py [hunk 7]
```diff
@@ -159,7 +159,8 @@ def get_fit_workspace_and_indices(fit):
 
             workspaces.append(workspace_name)
             indices.append(first_fit_index)
-            workspaces.append(workspace_name)
-            indices.append(second_fit_index)
+            if plot_diff:
```
Comment (StephenSmith25) [inline scripts/Muon/GUI/Common/plot_widget/plot_widget_model.py:162]: If you change the argument above you'll have to change it here as well.

**Review by StephenSmith25 — COMMENTED** [review]
Functions well, good work. Just a few comments.

**Review by StephenSmith25 — DISMISSED** [review]
Looks good to go 👍

**Review by StephenSmith25 — APPROVED** [review]

**Pull request merged**